modules/light-clients/07-tendermint/light_client_module_test.go

```diff
@@ -137,3 +144,173 @@ func (suite *TendermintTestSuite) TestRecoverClient() {
+			"cannot parse malformed client ID",
+			func() {
+				clientID = ibctesting.InvalidID
+			},
+			host.ErrInvalidID,
+		},
+		{
+			"client type is not 07-tendermint",
+			func() {
+				clientID = solomachineClientID
+			},
+			clienttypes.ErrInvalidClientType,
+		},
+		{
+			"cannot find client state",
+			func() {
+				clientID = tmClientID
+			},
+			clienttypes.ErrClientNotFound,
+		},
+		{
+			"upgraded client state is not for tendermint client state",
+			func() {
+				upgradedClientStateBz = []byte{}
+			},
+			clienttypes.ErrInvalidClient,
+		},
+		{
+			"upgraded consensus state is not tendermint consensus state",
+			func() {
+				upgradedConsensusStateBz = []byte{}
+			},
+			clienttypes.ErrInvalidConsensus,
+		},
+		{
+			"upgraded client state height is not greater than current height",
+			func() {
+				// last Height is at next block
+				lastHeight := clienttypes.NewHeight(1, uint64(suite.chainB.GetContext().BlockHeight()+1))
+
+				// zero custom fields and store in upgrade store
+				zeroedUpgradedClient := upgradedClientState.(*ibctm.ClientState).ZeroCustomFields()
+				zeroedUpgradedClientBz := clienttypes.MustMarshalClientState(suite.chainA.App.AppCodec(), zeroedUpgradedClient)
+				err := suite.chainB.GetSimApp().UpgradeKeeper.SetUpgradedClient(suite.chainB.GetContext(), int64(lastHeight.GetRevisionHeight()), zeroedUpgradedClientBz)
+				suite.Require().NoError(err)
+				err = suite.chainB.GetSimApp().UpgradeKeeper.SetUpgradedConsensusState(suite.chainB.GetContext(), int64(lastHeight.GetRevisionHeight()), upgradedConsensusStateBz)
+				suite.Require().NoError(err)
+
+				// change upgraded client state height to be lower than current client state height
+				clientState := path.EndpointA.GetClientState().(*ibctm.ClientState)
+				tmClient := upgradedClientState.(*ibctm.ClientState)
+				newLatestheight, ok := clientState.GetLatestHeight().Decrement()
+				suite.Require().True(ok)
+				tmClient.LatestHeight = newLatestheight.(clienttypes.Height)
+				upgradedClientStateBz = clienttypes.MustMarshalClientState(suite.chainA.App.AppCodec(), tmClient)
+
+				suite.coordinator.CommitBlock(suite.chainB)
+				err = path.EndpointA.UpdateClient()
+				suite.Require().NoError(err)
+
+				cs, found := suite.chainA.App.GetIBCKeeper().ClientKeeper.GetClientState(suite.chainA.GetContext(), path.EndpointA.ClientID)
+				suite.Require().True(found)
+
+				upgradedClientStateProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedClientKey(int64(lastHeight.GetRevisionHeight())), cs.GetLatestHeight().GetRevisionHeight())
+				upgradedConsensusStateProof, _ = suite.chainB.QueryUpgradeProof(upgradetypes.UpgradedConsStateKey(int64(lastHeight.GetRevisionHeight())), cs.GetLatestHeight().GetRevisionHeight())
+			},
+			ibcerrors.ErrInvalidHeight,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		suite.Run(tc.name, func() {
+			suite.SetupTest() // reset
+			cdc := suite.chainA.App.AppCodec()
+			ctx := suite.chainA.GetContext()
```

Comment:
super nit: one thing I dislike about assigning the context to a variable rather than dynamically requesting it is that this variable has the incorrect block height set in the context. The `tc.malleate()` call will commit a block on chainA on the "success" case, resulting in the context having an out of date block header when calling `VerifyUpgradeAndUpdateState`. I guess it is fine in this test, but can lead to confusing errors if it needs to be accurate

Reply:
Got it. I changed it now.